aws-lc-rs/src/kem.rs

```diff
@@ -0,0 +1,530 @@
+        }?;
+        let privkey = LcPtr::new(unsafe {
+            EVP_PKEY_kem_new_raw_secret_key(alg.id.nid(), bytes.as_ptr(), bytes.len())
+        })?;
+        Ok(DecapsulationKey {
+            algorithm: alg,
+            evp_pkey: privkey,
+            priv_key: bytes.into(),
+        })
+    }
+}
+
+unsafe impl<Id> Send for DecapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+unsafe impl<Id> Sync for DecapsulationKey<Id> where Id: AlgorithmIdentifier {}
+
+impl<Id> Drop for DecapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn drop(&mut self) {
+        self.priv_key.zeroize();
+    }
+}
+
+impl<Id> AsRef<[u8]> for DecapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn as_ref(&self) -> &[u8] {
+        &self.priv_key
+    }
+}
+
+impl<Id> Debug for DecapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PrivateKey")
+            .field("algorithm", &self.algorithm)
+            .finish_non_exhaustive()
+    }
+}
+
+/// A serializable public key usable with KEMS. This can be constructed
+/// from a `KemPrivateKey` or constructed from raw bytes.
+pub struct EncapsulationKey<Id = AlgorithmId>
+where
+    Id: AlgorithmIdentifier,
+{
+    algorithm: &'static Algorithm<Id>,
+    evp_pkey: LcPtr<EVP_PKEY>,
+    pub_key: Box<[u8]>,
+}
+
+impl<Id> EncapsulationKey<Id>
+where
+    Id: AlgorithmIdentifier,
+{
+    /// Return the algorithm associated with the given KEM public key.
+    #[must_use]
+    pub fn algorithm(&self) -> &'static Algorithm<Id> {
+        self.algorithm
+    }
+
+    /// Performs the encapsulate operation using the current KEM public key, generating a ciphertext
+    /// and associated shared secret.
+    ///
+    /// After the encapsulation is finished, `encapsulate` calls `kdf` with the ciphertext and raw shared secret
+    /// from the operation and then returns what `kdf` returns.
+    ///
+    /// # Errors
+    /// `error::Unspecified` when operation fails due to internal error.
+    pub fn encapsulate(&self) -> Result<(Ciphertext<'static>, SharedSecret), Unspecified> {
+        let mut ciphertext_len = self.algorithm.ciphertext_size();
+        let mut shared_secret_len = self.algorithm.shared_secret_size();
+        let mut ciphertext: Vec<u8> = vec![0u8; ciphertext_len];
+        let mut shared_secret: Vec<u8> = vec![0u8; shared_secret_len];
+
+        let ctx = LcPtr::new(unsafe { EVP_PKEY_CTX_new(*self.evp_pkey, null_mut()) })?;
+
+        if 1 != unsafe {
+            EVP_PKEY_encapsulate(
+                *ctx,
+                ciphertext.as_mut_ptr(),
+                &mut ciphertext_len,
+                shared_secret.as_mut_ptr(),
+                &mut shared_secret_len,
+            )
+        } {
+            return Err(Unspecified);
+        }
+
+        ciphertext.truncate(ciphertext_len);
+        shared_secret.truncate(shared_secret_len);
```

Comment:
why is truncation needed here?

Reply:
This is just safety step to ensure that AWS-LC didn't return us a smaller length then value then the buffer we passed in. Right now the buffer is sized to the expected size so it should be a no-op.